Modify history behaviour, DG and UG fixes

Lysire opened the pull request:
### What this does
This PR changes the behaviour of the `history` command. Previously, the command listed previously entered commands in reverse chronological order. However, it now lists previously entered commands in chronological order, along with a positive integer before each command which indicates its position in the execution sequence (e.g. `3    clear` means that the command `clear` was the 3<sup>rd</sup> to be executed). Meanwhile, the displaying of commands in the `CommandBox` remain unaltered (only the commands themselves are listed). The reason for this change is to allow for a more detailed listing of previously entered commands which may provide to be more intuitive, resembling the output provided by the bash history command. 

The PR also updates the UG, DG as well as fix some failing tests based on the above change in behaviour. Additionally, the dependency on `KeyboardShortcutsEnum` for `CommandBox` is reflected as well. 

<!-- Delete accordingly -->

### How to test
<!-- Write an essay here if it's more suitable -->
1. Launch the application.
2. Type in a few commands
3. Ensure that the result output by `history` is in the format `n    command` where `n` is an integer.

### Notes
<!-- Anything else we should take note of, e.g. how this affects future PRs,
issues faced, etc. -->
N/A

Lysire commented:
As a side note, I realize that the package definition for `EditRoomCommandParser` was not quite right so I made sure to fix this, along with a few minor errors related to this erroneous package definition (e.g. import statements in `CommandWordEnum`).

Lysire commented:
Update: 
1. Add class diagram for undo/redo
2. Add activity diagram for `undo` command
3. Remove mentions of `Person` in UG and DG.

mkeoliya commented on docs/DeveloperGuide.md:
```diff
@@ -86,7 +86,7 @@ The sections below give more details of each component.
 [`Ui.java`](https://github.com/se-edu/addressbook-level3/tree/master/src/main/java/seedu/address/ui/Ui.java)
 
-The UI consists of a `MainWindow` that is made up of parts e.g.`CommandBox`, `ResultDisplay`, `PersonListPanel`, `StatusBarFooter` etc. All these, including the `MainWindow`, inherit from the abstract `UiPart` class.
+The UI consists of a `MainWindow` that is made up of parts e.g.`CommandBox`, `ResultDisplay`, `StudentListPanel`, `StatusBarFooter` etc. All these, including the `MainWindow`, inherit from the abstract `UiPart` class.
```
Good catch here!

mkeoliya commented on docs/DeveloperGuide.md:
```diff
@@ -271,25 +271,33 @@ the `Allocation` association class (alternative 4).
-- `VersionedResiReg#save()` — Saves the current residence regulation state in its history.
-- `VersionedResiReg#undo()` — Restores the previous residence regulation state from its history.
-- `VersionedResiReg#redo()` — Restores a previously undone residence regulation state from its history.
+- `StatefulResiReg#save()` — Saves the current residence regulation state in its history.
```
I think we can shorten `residence regulation` here to `ResiReg`

Lysire replied:
Resolved. I've replaced every instance of "residence regulation" with "ResiReg".

mkeoliya commented on src/main/java/seedu/resireg/logic/commands/HistoryCommand.java:
```diff
@@ -3,37 +3,44 @@
-        ArrayList<String> previousCommands = new ArrayList<>(history.getHistory());
 
-        if (previousCommands.size() == 0) {
+        ArrayList<String> commands = new ArrayList<>(history.getHistory());
+        Integer numCommands = history.getCounter();
+
+        for (int i = 1; i <= numCommands; i++) {
+            String labelled = i + "\t" + commands.get(i - 1);
+            commands.set(i - 1, labelled);
+        }
+
+        if (commands.isEmpty()) {
```
This should be refactored to a method inside CommandHistory, probably called `getResultAsString`. Maintains SLAP and prevents redundant getter methods.

Lysire replied:
Resolved. Thanks for pointing this out. I've extracted the above code fragment into a method called `getCommandResultString()` in `CommandHistory` to adhere to SLAP. I've also added a test for this method.

mkeoliya commented on src/main/java/seedu/resireg/ui/CommandBox.java:
```diff
@@ -42,9 +44,9 @@
-        this.history = history;
         // calls #setStyleToDefault() whenever there is a change to the text of the command box.
         commandTextField.textProperty().addListener((unused1, unused2, unused3) -> setStyleToDefault());
+        this.history = history;
```
What's the rationale behind changing the order of execution?

Lysire replied:
Just a small edit, it was to group the initalization of the history related fields together.

mkeoliya commented on src/test/java/seedu/resireg/logic/CommandHistoryTest.java:
```diff
@@ -45,6 +45,15 @@ void getHistory() {
 
+    @Test
+    void getCounter() {
+        final CommandHistory historyWithAC = new CommandHistory();
```
The `historyWithAC` object probably be defined as `VALID_COMMAND_HISTORY` in `CommandTestUtils`, so that it can be reused by other tests, and to maintain consistency with other testing.

Lysire replied:
Thanks for pointing this out, but I don't think not defining the objects inside `CommandTestUtils` will cause inconsistency, since such testing is done by classes under the Model package. That said, I've used the code fragment annotated with `@BeforeEach` to remove duplicate initialization statements.

mkeoliya commented on src/test/java/seedu/resireg/logic/LogicManagerTest.java:
```diff
@@ -162,9 +164,16 @@ private void assertCommandFailure(String inputCommand, Class<? extends Throwable
         try {
             CommandResult result = logic.execute(HistoryCommand.COMMAND_WORD);
+
+            List<String> expected = Arrays.asList(expectedCommands);
```
Good to refactor this to `getResultAsString()` as well.

Lysire replied:
Resolved. Again, I've extracted the above code fragment into a private method called getExpectedResultString() to adhere to SLAP.

mkeoliya commented on src/test/java/seedu/resireg/logic/commands/HistoryCommandTest.java:
```diff
@@ -14,22 +14,27 @@
     private Model model = new ModelManager();
     private Model expectedModel = new ModelManager();
 
+    private final String labelOne = "1" + "\t";
+    private final String labelTwo = "2" + "\t";
+    private final String labelThree = "3" + "\t";
```
As above, these labels should be placed in a constants file. `TypicalCommandHistories` works as well, especially if we're looking to add a `CommandHistoryBuilder` or add integration tests in the future.

Lysire replied:
In line with what has been mentioned above, I've placed the constants in a `TypicalPositions` file.

mkeoliya reviewed (commented):
Left a few comments to improve code and testing quality.

mkeoliya commented on src/main/java/seedu/resireg/logic/CommandHistory.java:
```diff
@@ -53,11 +69,12 @@ public boolean equals(Object o) {
         CommandHistory other = (CommandHistory) o;
-        return history.equals(other.history);
+        return history.equals(other.history)
+                && counter.equals(other.counter);
```
Lacks test coverage. (Probably add a test where there's two `CommandHistory` objects with the same `history` but different `counter`). Useful if for some reason the list of command histories gets modified unsafely.

Lysire replied:
Resolved.

Lysire commented:
Update: To further resemble the output provided by the bash history command, I've made changes to the behaviour of the `history` command to only add nonempty commands (that is simply pressing <kbd>enter</kbd> will not be reflected in the history). I've updated the UG and DG to reflect these changes. 

With regard to the UG and DG updates, I've added "state" in front of "history" for the undo/redo implementation section to clearly demarcate that the state history is recorded in the case of undo/redo.

jetnew reviewed (approved):
LGTM!
